### DISCUSSION TITLE
Pi5 at full power with Create3?

### DISCUSSION THREAD

**Discussion by slowrunner** [iRobotEducation/create3_docs#443] [Questions] [answered]
### How are you connecting to your Create 3?

Wi-Fi (ROS 2)

### Computer(s) Model(s) and Operating System(s)

Desired: Raspberry Pi 5 (with Bookworm)

### Which version of ROS 2 is installed on your computer?

Humble

### Which firmware version is installed on your robot?

x.x

### Which RMW is your robot running?

CycloneDDS

### Is the robot connected to a network? If so, what is the network type?

home network

### Are there multiple Create 3 robots connected to your network?

No

### Is multicast enabled?

None

### What is the Adapter Board's USB/BLE Toggle currently switched to?

None

### Describe your question.

The Create3 supports 5V 3A power delivery over the USB-C connector which is sufficient for Raspberry Pi 4, but will not support the new Raspberry Pi 5 at full speed which claims to need 5V at 5A (25W) for full 2.5GHz processing.

Is there a USB-C powerbank that could charge from the Create3 and simultaneously support periods of maximum load needed by the Raspberry Pi 5 (and perhaps the Oak-D-Lite camera as well)?  

Raspberry Pi 5 <-- 25W PD Powerbank <-- iRobot_Create3 

Would this perhaps also extend the off dock run time?

(The Pi5 claims to offer lower power needed for Pi4 performance, so other than the disadvantage of needing a Bookworm level OS, there could be some possible advantage in run-time for the direct mating solution as well with careful load management.)

[Something like this would work](https://www.amazon.com/dp/B08THCNNCS/)?

**Comment by brianabouchard** [comment]
I've been exploring how this would work as well. From the release statement: https://www.raspberrypi.com/news/introducing-raspberry-pi-5/ it sounds like we should be able to run the Pi 5 at 15 W with the primary impact being the limit on downstream USB current. However, they then go on to say: 
"It should be noted that users have the option to override the current limit, specifying the higher value even when using a 3A adapter. In our testing, we have found that in this mode Raspberry Pi 5 functions perfectly well with typical configurations of higher-power USB devices, and all but the most pathological workloads."

Once mine arrives, I'll be curious to see the limits to which we can push them within the bounds of the 15 W supply from the Create 3.

**Reply by slowrunner** [reply]
My Pi5 arrives next week, and I'm still wondering if the Create3 is going to support the Pi5 and sensors?  

My present GoPiGo3 based ROS2bot is pegging a Pi4 at 100% 73C and only getting 45-75 minutes "playtime", so I want to build my next-gen bot around the Pi5, and the YDLIDAR X4, and Oak-D-light from my current bot.

I briefly had a TurtleBot4lite and it seemed that configuration/battery capacity was only going to get an hour playtime with a Pi4, LIDAR, and Oak-D-Lite.  I have the feeling, I need to design a piggyback 30-50Wh power supply that will recharge from the Create3 battery to supply the processor and sensors with the Create3 battery only supplying what the Create3 needs when off the dock.

**Reply by brianabouchard** [reply]
Good news, I have a small update on this. My Pi 5 arrived recently, but I have only been able to do very preliminary testing. I can confirm that the Create 3 robot can power the Pi 5, but as expected, not at the full 25W. In order to get it to boot, I had to add ` usb_max_current_enable=1` to the `config.txt` which, according to the docs, allows it to run on a 3A power supply. 

The biggest challenge I'm running into at the moment is getting the Pi to recognize the usb0 connection, but as I said, I haven't had much time for testing so this could be a minor challenge once I have more time. I'll keep you posted on what I learn!

**Reply by slowrunner** [reply]
Cool!  

I ran some perf tests:

# MY Pi5 PERFORMANCE TESTS

Raspberry Pi5 8GB mounted in official Pi5 case with fan,  
and Pi5 4GB in case with cooler  

- Python3 floating point addition 0.000973ms - 3.5x faster than Pi4  
- Python3 integer addition 0.000657ms - 3.8x faster than Pi4  
- Four CPU Stress for 5 minutes: No throttling, 70.3C temp  
  Same max temp with cooling fan case, and Pi5 cooler in case  
  7.2W 1.4A during full stress  
  (3W 0.6A idle)  

Perhaps the Create3 will even support full load if the auxiliary sensors are not powered from the Pi5 USB (e.g. via a step-down 5v supply from the aux power jack of the Create3)

BTW - I documented setting up ROS 2 Humble in Docker on the Pi5 (PiOS Bookworm, but same would work for Ubuntu 23)

https://github.com/slowrunner/pi5desk/tree/main/configs/docker

OH COOL! I just found an "Official Desktop Create3 Humble Docker" image:
https://github.com/iRobotEducation/create3-docker
This is really encouraging.

**Comment by slowrunner** [comment, marked as answer]
I just did a 100% Pi5 load test powered from the Create3 USB-C port (no USB peripherals powered by the Pi5) 

NO PROBLEM!  Full Speed, Full Load, No Throttling - 12.7W total
- Create3: 3.44W
- Pi5: 8.46W
- Audio Speaker (0.8W idle from 14.4v aux powered hub) 
- No driving



```
voltage: 15.54800033569336
temperature: 32.64999771118164
current: -0.8140000104904175
charge: 1.6330000162124634
capacity: 1.9220000505447388
design_capacity: 1.9220000505447388
percentage: 0.8500000238418579


12.7W max  

********** Pi5DESK MONITOR ******************************
Saturday 12/16/23
 18:15:51 up 18 min,  4 users,  load average: 4.90, 3.15, 1.64
temp=69.2'C
frequency(0)=2400008192
throttled=0x0
               total        used        free      shared  buff/cache   available
Mem:           7.9Gi       562Mi       6.5Gi        33Mi       928Mi       7.3Gi
Swap:           99Mi          0B        99Mi

```

I'm going to mark this question as answered 🎆

**Comment by slowrunner** [comment]
Interestly at 25% load, the powered hub is supplying roughly 300mA to the Pi5, with the USB C connector supplying another 300mA to the Pi5 for the total 3W draw.

https://forums.raspberrypi.com/viewtopic.php?t=361704

**Reply by slowrunner** [reply]
Further testing and consensus of the vociferous netizens suggests "the hub is garbage and unsafe to use".
SO... I dispensed with the hub and plugged my speaker and Oak-D-Lite directly into the Pi5 and fired it all up!  (Nothing smoked.)

Running the Oak-D-Lite depthai_demo draws 3.4W from the USB 3 port of the Pi5, and the processor load was 60%, for a total draw from the Create3 for the RPi5, Oak-D-Lite running, and an audio speaker was ...

ta da!  

9.5W  leaving some room for the TBD: ROS 2 Humble in Docker load. 
![Screenshot 2023-12-18 at 11 07 10 AM](https://github.com/iRobotEducation/create3_docs/assets/8950202/8ae1b3a7-f297-4a4b-85bf-7598d54ec78a)

**Reply by brianabouchard** [reply]
This is super interesting! I'm glad you're able to run the Oak-D-Lite right off of the Pi5's USB ports. This is great info to have and share with the community!

**Reply by slowrunner** [reply]
> Interestly at 25% load, the powered hub is supplying roughly 300mA to the Pi5, with the USB C connector supplying another 300mA to the Pi5 for the total 3W draw.

I tried a "USB Power Blocker" and the "back power" is greatly reduced, but not eliminated.  This is totally incomprehensible to me how power is flowing from the 14.4v battery through the hub to the Raspberry Pi 5 USB port somehow along the ground wire.

In Wali's current configuration (no LIDAR) he has enough power via the USB-C connector to drive the Oak-D-Lite, the speaker, and the USB Wireless gamepad adapter, with 100% processor load.  

Therefore I don't need to solve this issue, but if I ever decide to add a LIDAR, it will be at the cost of processing load, or I will have to add a step-down 5v power-buck to power the LIDAR.  If I ever want to add a USB SSD drive, the powered hub configuration will become necessary to solve.  (The backpowering didn't burn my processor, but it really is not a good idea.)

**Comment by slowrunner** [comment]
WaLi is very happy running his Raspberry Pi 5 "brain" and has not run into any throttling or USB power issues.

Closing this as "Yes, Pi5 at full power with Create3 - no problem"

```
*** Create3-WALI TOTAL LIFE STATISTICS ***
Total Awake:  434.68  hrs
Total Naps:     24.85  hrs
Total Life:  459.53 hrs (since Dec 11, 2023)
Playtimes (Undocked-Docked): 59
Sessions (boot):  25
Average Session:  18.3 hrs
Safety Shutdowns:  3

```

**Reply by brianabouchard** [reply]
Awesome! @slowrunner as a follow up, were you able to successfully get the connection working over the usb0 interface or are you using Wi-Fi?

**Comment by slowrunner** [comment]
@brianabouchard 

>  were you able to successfully get the connection working over the usb0 interface or are you using Wi-Fi?

I am using WiFi for a few reasons:
- simpler to configure
- I can "nap" the Pi5 during the night and the Create3 remains able to see ntp
- I can view http://create3-wali.local/about and logs/,  and do firmware updates without the Pi5 active or in the middle

When Wali is much more autonomous, and has greater safety monitoring / emergency response abilities, I will want to eliminate the dependence on WiFi between the Create3 and the Pi5, but for now I don't know of any USB0 networking advantages blocking his "search for intelligence"  😄 

```
*** Create3-WALI TOTAL LIFE STATISTICS ***
Total Awake:  1024.83  hrs
Total Naps:     113.3  hrs
Total Life:  1138.13 hrs (since Dec 11, 2023)
Playtimes (Undocked-Docked): 105
Sessions (boot):  35
Average Session:  29.2 hrs
Safety Shutdowns:  3
Total Travel:  732.3 meters 2402.5 feet

Recent dead-reconing trip to front door to check the locks (except camera not implemented yet):

2024-01-10 07:58|wali_node.py| ** WaLI Undocking: success at battery 100% **
2024-01-10 08:37|[logMaintenance.py.main]** Dead-Reckoning To The Front Door From Undocked (and back) **
2024-01-10 08:40|[logMaintenance.py.main]** Arrived at front door **
2024-01-10 08:40|[logMaintenance.py.main]** Dead-Reckoning Back To Undocked Position From The Front Door **
2024-01-10 08:46|[logMaintenance.py.main]** Manual Docking: success (assumed) **
2024-01-10 08:47|[logMaintenance.py.main]** Issued reset_pose **
2024-01-10 09:15|wali_node.py| ** WaLI Undocking: success at battery 100% **
2024-01-10 12:24|wali_node.py| ** WaLI Ready To Dock at battery 17% **
2024-01-10 12:31|wali_node.py| ** WaLI dock goal result - Docking: success at battery 14% **
2024-01-10 13:24|lifelogger.dEmain execution: 23.45

```
